langchain/src/vectorstores/typeorm.ts

```diff
@@ -0,0 +1,307 @@
+          sourceType: doc.sourceType,
+        },
+      });
+
+      const documentsThatWillBeUpserted = rows.filter(
+        (d) => d.sourceName === doc.sourceName
+      );
+      const idsDocumentsThatWillBeUpserted = documentsThatWillBeUpserted.map(
+        (d) => d.id
+      );
+
+      const idsDocumentsThatWeShouldDeleted = documentsInDatabase
+        .filter((d) => !idsDocumentsThatWillBeUpserted.includes(d.id))
+        .map((d) => d.id);
+
+      this.verboseConsoleLog(
+        `Ids has changed for document ${doc.sourceName} and ${doc.sourceType}, action: delete this ids: ${idsDocumentsThatWeShouldDeleted}`
+      );
+
+      await documentRepository.delete({
+        id: {
+          in: idsDocumentsThatWeShouldDeleted,
+        },
+      });
+    }
+
+    const documentsToUpsert = [];
+    for (const row of rows) {
+      // Search for existing document with same sourceName and sourcetype
+      const documentsInDatabase = (await documentRepository.findOne({
+        where: {
+          id: row.id,
+        },
+      })) as Document;
+
+      if (documentsInDatabase) {
+        if (documentsInDatabase.hash === row.hash) {
+          this.verboseConsoleLog(
+            `Same hash was found for id ${row.id}, action: ignore`
+          );
+        } else {
+          documentsToUpsert.push(row);
+          this.verboseConsoleLog(
+            `Different hash was found for id ${row.id}, action: update embedding`
+          );
+        }
+      } else {
+        documentsToUpsert.push(row);
+        this.verboseConsoleLog(
+          `No document was found for id ${row.id}, action: create`
+        );
+      }
+    }
+
+    const chunkSize = 500;
+    for (let i = 0; i < documentsToUpsert.length; i += chunkSize) {
+      const chunk = documentsToUpsert.slice(i, i + chunkSize);
+
+      try {
+        await documentRepository.upsert(chunk, [`id`]);
+      } catch (e) {
+        console.error(e);
+        throw new Error(`Error inserting: ${chunk[0].pageContent}`);
+      }
+    }
+  }
+
+  async similaritySearchVectorWithScore(
+    query: number[],
+    k: number,
+    filter?: this["FilterType"]
+  ): Promise<[TypeORMVectorStoreDocument, number][]> {
+    const embeddingString = `'[${query.join(",")}]'`;
+    const _filter = filter ?? "'{}'";
+    const queryString = `
+      SELECT *, embedding <=> ${embeddingString}::vector as "_distance" 
+      FROM ${this.tableName}
+      WHERE metadata @> ${_filter}
+      ORDER BY "_distance" ASC
+      LIMIT ${k};
+    `;
+    const documents = await this.appDataSource.query(queryString);
+
+    const results = [] as [TypeORMVectorStoreDocument, number][];
+    for (const doc of documents) {
+      if (doc._distance != null && doc.pageContent != null) {
+        const document = new Document(doc) as TypeORMVectorStoreDocument;
+        document.id = doc.id;
+        results.push([document, doc._distance]);
+      }
+    }
+
+    return results;
+  }
+
+  async initTable(): Promise<void> {
```

Comment:
was wondering if we should call this function at the beginning of fromDocuments function to be sure the table and the extension exist as it doesn't cost much in time.

Reply:
done

Reply:
You can call this `ensureTable` which might be a bit more clearly named.